core/bin/external_node/src/main.rs

```diff
@@ -210,42 +210,37 @@ async fn run_core(
     .await?;
 
     task_handles.push(tokio::spawn({
-        let ctx = ctx::root();
-        let cfg = config.consensus.clone();
-        let mut stop_receiver = stop_receiver.clone();
-        let fetcher = consensus::Fetcher {
-            store: consensus::Store(connection_pool.clone()),
-            sync_state: sync_state.clone(),
-            client: Box::new(main_node_client.clone()),
-            limiter: limiter::Limiter::new(
-                &ctx,
-                limiter::Rate {
-                    burst: 10,
-                    refresh: time::Duration::milliseconds(30),
-                },
-            ),
+        let config = config.consensus.clone();
+        let secrets =
+            config::read_consensus_secrets().context("config::read_consensus_secrets()")?;
+        let cfg = match (config, secrets) {
+            (Some(cfg), Some(secrets)) => Some((cfg, secrets)),
+            (Some(_), None) => {
+                anyhow::bail!("Consensus config is specified, but secrets are missing")
+            }
+            (None, _) => {
+                // Secrets may be unconditionally embedded in some environments, but they are unused
+                // unless a consensus config is provided.
+                None
+            }
         };
-        let actions = action_queue_sender;
+
+        let pool = connection_pool.clone();
+        let sync_state = sync_state.clone();
+        let main_node_client = Arc::new(main_node_client.clone());
+        let stop_receiver = stop_receiver.clone();
         async move {
-            scope::run!(&ctx, |ctx, s| async {
-                s.spawn_bg(async {
-                    let res = match cfg {
-                        Some(cfg) => {
-                            let secrets = config::read_consensus_secrets()
-                                .context("config::read_consensus_secrets()")?
-                                .context("consensus secrets missing")?;
-                            fetcher.run_p2p(ctx, actions, cfg.p2p(&secrets)?).await
-                        }
-                        None => fetcher.run_centralized(ctx, actions).await,
-                    };
-                    tracing::info!("Consensus actor stopped");
-                    res
-                });
-                ctx.wait(stop_receiver.wait_for(|stop| *stop)).await??;
-                Ok(())
-            })
+            let ctx = ctx::root();
+            consensus::run_fetcher(
+                &ctx,
+                cfg,
+                pool,
+                sync_state,
+                main_node_client,
+                action_queue_sender,
+                stop_receiver,
+            )
```

Comment:
as I see it, you wrap Fetcher{..}.run() with run_fetcher(...) which just obscures which argument does what. What's the purpose?

Reply:
This is a sort of "default entrypoint". For some time now we'll have to maintain both framework and external node. If we'll have a copy-pasted implementation, it may run out of sync. With a single implementation, change one place does the thing for both EN and the framework.

All the public APIs are still available in the module, so if anything custom is required (though I doubt it will anytime soon), user is free to do so.